openshift/osd-network-verifier
# Provide CA Certificates via cloud-init's ca_certs module

- Bug

## What does this PR do? / Related Issues / Jira

Currently, the curl flag we're using to provide a ca cert will _only_ use the provided cert - we want to use both the supplied SSL cert and RHEL's OEM trust bundle. This change moves away from the curl flag and instead uses the `ca_certs` module of cloud-init.

Resolves [OSD-24353](https://issues.redhat.com//browse/OSD-24353)

## Checklist


## How to test this PR locally / Special Instructions
See [OSD-24353](https://issues.redhat.com//browse/OSD-24353) for setting up a transparent proxy. I used https://gitlab.cee.redhat.com/travi/osd-proxy-infra for setting up a non-transparent proxy.

I ran these new changes against
- a transparent proxy
- a non-transparent proxy
- a split-cert transparent proxy

In all situations, I saw the expected results. Elaborating more on the split-cert situation, I set up a transparent proxy using mitmproxy with `--ignore-hosts registry.redhat.io:443` in the service file. When I ran network verifier without passing a capath, _all_ URLs failed *_EXCEPT_* registry.redhat.io

## Logs 

With this change, the userdata template expands to something like the following:
```
#cloud-config

ca_certs:
  trusted:
    - |-
      -----BEGIN CERTIFICATE-----
      THE CERTIFICATE
      -----END CERTIFICATE-----
runcmd:
  - systemctl mask --now serial-getty@ttyS0.service
  - dmesg -D
  - echo "NV_CURLJSON_BEGIN" >/dev/ttyS0
  - export http_proxy= https_proxy=
  - curl --capath /etc/pki/tls/certs/ --retry 3 --retry-connrefused -t B -Z -s -I -m ${TIMEOUT} -w "%{stderr}${LINE_PREFIX}%{json}\n" ${CURLOPT} ${URLS} --proto =http,https,telnet ${TLSDISABLED_URLS_RENDERED} 2>/dev/ttyS0
  - echo "NV_CURLJSON_END" >/dev/ttyS0
power_state:
  delay: 2
  mode: poweroff
  message: Auto-terminating instance due to timeout
  timeout: 300
```
